File: solutions/day14/main.cpp
#include <vector>
#include <cstdint>
#include <array>
#include <algorithm>
#include <iostream>

struct board
{
    std::vector<uint8_t> buf;
    std::array<size_t, 2> elves;
};

void do_step(board&);

std::string part1(board, size_t steps);
size_t part2(board, std::string_view sought);

int main()
{
    board b {
        {3,7,1,0},
        {0, 1}        
    };

    std::cout << "Part 1: " << part1(b, 505961) << '\n';
    std::cout << "Part 2: " << part2(b, "505961") << '\n';
}

std::string part1(board b, size_t steps)
{
    while(b.buf.size() < steps + 10) {
        do_step(b);
    }
    std::string ret(10,0);
    std::transform(b.buf.begin()+steps, b.buf.begin()+steps+10, ret.begin(),
        [](uint8_t x) -> char { return '0' + x; });
    return ret;
}

size_t part2(board b, std::string_view sought)
{
    std::vector<uint8_t> cmp(sought.size());
    std::transform(sought.begin(), sought.end(), cmp.begin(),
        [](char c) -> uint8_t { return c - '0'; });

    b.buf.reserve(200'000'000);

    while(true) {
        size_t prev_size = b.buf.size();
        do_step(b);
        auto diff = b.buf.size() - prev_size;

        for(size_t i = 0; i < diff; i++) {
            if(b.buf.size() > sought.size() + i) {
                if(std::equal(
                    cmp.rbegin(), cmp.rend(),
                    b.buf.rbegin() + i, b.buf.rbegin() + sought.size() + i)){
                    return b.buf.size() - sought.size() - i;
                }
            }
        }
    }
}

void do_step(board& board)
{
    auto new_value = board.buf[board.elves[0]] + board.buf[board.elves[1]];
    
    if(new_value > 9) {
        board.buf.push_back(new_value / 10);
        board.buf.push_back(new_value % 10);
    } else {
        board.buf.push_back(new_value);
    }

    for(auto& e : board.elves) {
        e = board.buf[e] + 1 + e;
        e %= board.buf.size();
    }
}

File: solutions/day14/rust/src/main.rs
use std::io;

#[derive(Debug, Clone)]
struct Board
{
    buf: Vec<u8>,
    elves: [usize; 2],
}

fn main() -> Result<(), io::Error>
{
    let board = Board{ buf: vec!(3, 7, 1, 0), elves: [0, 1] };
    let input = 505961;
    println!("Part 1: {}", part1(input, board.clone()));
    println!("Part 2: {}", part2(input.to_string(), board));
    Ok(())
}

fn part1(steps: usize, mut b: Board) -> String
{
    while b.buf.len() < steps + 10 {
        do_step(&mut b)
    }
    let len = b.buf.len();

    b.buf[len-10..len].iter()
        .map(|c| (c + 0x30) as char)
        .collect()
}

fn part2(sought_str: String, mut b: Board) -> usize
{
    let sought: Vec<u8> = sought_str
        .chars()
        .map(|c| c as u8 - 0x30)
        .collect();

    b.buf.reserve(200_000_000);

    loop {
        let prev_size = b.buf.len();
        do_step(&mut b);
        let new_size = b.buf.len();
        let size_diff = new_size - prev_size;
        for i in 0..size_diff {
            if i + sought.len() > new_size {
                break
            }

            let rng = &b.buf[new_size-i-sought.len()..new_size-i];
            if rng == &sought[..] {
                return new_size - sought.len() - i;
            }
        }
    }
}

fn do_step(b: &mut Board) -> ()
{
    let new_val : u8 = b.elves
        .iter()
        .map(|e| b.buf[*e])
        .sum();

    if new_val > 9 {
        b.buf.push(new_val / 10);
    }
    b.buf.push(new_val % 10);

    for e in b.elves.iter_mut() {
        let nv = b.buf[*e] as usize + 1 + *e;
        let nv = nv % b.buf.len();
        *e = nv;
    }
}
